Migrate MachineLearningServices-preview-10-01 to TypeSpec

mcgallan opened the pull request:
This PR migrates your latest version (identified by the tag in your readme.md) of swagger to TypeSpec. We already tried our best to make sure the TypeSpec represents same as previous swagger. Since we lack the business knowledge, please validate this PR again to make sure it's functional equivalent as before. The local validation step is at [Getting started | TypeSpec Azure](https://azure.github.io/typespec-azure/docs/migrate-swagger/01-get-started/#review-and-adjust-the-typespec)

Besides, TypeSpec encourages to follow [ARM guidelines](https://aka.ms/rpguidelines).
Please reach out to [TypeSpec Discussions Channel](https://teams.microsoft.com/l/channel/19%3a906c1efbbec54dc8949ac736633e6bdf%40thread.skype/TypeSpec%2520Discussion%2520%25F0%259F%2590%25AE?groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47) if there is any help needed.

Compilation Issues Encountered
1. Path Parameter Pattern mismatch between operations on the same path
In MachineLearningServices, path parameter patterns for create operations often differ from get, update, or delete operations on the same path. If needed, I can list these operations and their path parameters, for example:
CapabilityHosts_CreateOrUpdate, CodeContainers_CreateOrUpdate, CodeVersions_CreateOrUpdate, ComponentContainers_CreateOrUpdate, ComponentVersions_CreateOrUpdate, DataContainers_CreateOrUpdate with parameter name
BatchDeployments_Update with parameters endpointName and deploymentName

2. Duplicate Enum Names with Different Values
EndpointType is used in both EndpointDeployment_GetInWorkspace and Endpoint_List operations in the original swagger with the same name but actually represents two different enums. TypeSpec doesn't support this approach. EndpointServiceConnectionStatus enum has the same issue.

3. Model-Level Secret Addition
AccountKeyDatastoreCredentials as a model cannot add secrets at the model level.

4. Example Addition Not Supported
TypeSpec cannot add examples.

5. Required Property Redefinition
In AzureOpenAiFineTuning and CustomModelFineTuning, there are cases where required model properties from parent models are set as required again (i.e., modelProvider). Similar situations exist with resource models like batchdeployment where location has the same issue. This causes breaking changes in JS, making it impossible to implement.

6. Format Override Issue
In SASCredentialDto, the sasUri format uri is overridden by secret to become format password.

7. ComputeResource Resource Type Conversion Issue
In the original swagger, ComputeResource was a ProxyResource with custom location and tags properties. During conversion, it was consolidated into a TrackedResource, which caused Original location and tags properties to gain additional "x-ms-mutability" attributes and The tags property lost its x-nullable attribute.

mcgallan committed:
convert

mcgallan committed:
update and compile

mcgallan committed:
update examples

mcgallan committed:
update

mcgallan committed:
update

mcgallan committed:
delete

mcgallan committed:
update cspell

mcgallan committed:
update finalresult

mcgallan committed:
update finalresult

mcgallan committed:
remove operationId and related suppression

mcgallan committed:
update tspcofig

mcgallan committed:
fix

mcgallan committed:
rename folder

mcgallan committed:
update

mcgallan committed:
delete

mcgallan committed:
remove private

mcgallan committed:
Merge branch 'main' into Convert-MachineLearningServices-preview-10-01

mcgallan committed:
remove csharp in tspconfig

mcgallan committed:
Merge remote-tracking branch 'upstream/main' into Convert-MachineLearningServices-preview-10-01

mcgallan committed:
Fix uniqueItems

mcgallan committed:
fix update

mcgallan committed:
Add Secret on model properties and Fix ComputeResource

mcgallan committed:
Merge branch 'main' into Convert-MachineLearningServices-preview-10-01

mcgallan committed:
update

mcgallan committed:
fix ts validation error

mcgallan committed:
Merge remote-tracking branch 'upstream/main' into Convert-MachineLearningServices-preview-10-01

mcgallan committed:
regen

mcgallan committed:
remove hierarchyBuilding

mcgallan committed:
Merge branch 'main' into Convert-MachineLearningServices-preview-10-01

v-jiaodi committed:
Merge branch 'main' into Convert-MachineLearningServices-preview-10-01

weidongxu-microsoft commented on specification/machinelearningservices/MachineLearningServices.Management/Workspace.tsp:
```diff
@@ -0,0 +1,313 @@
+    void,
+    Response = {
+      @body body: void;
+    }
```
You need to add `FinalResult = ` to these LRO ops.

Other places

```
08:04:36.616 cmderr W 	[Java-Sdk-Automation] /mnt/vss/_work/1/s/azure-sdk-for-java/sdk/machinelearningservices/azure-resourcemanager-machinelearningservices/TempTypeSpecFiles/MachineLearningServices.Management/Workspace.tsp:226:16 - error @azure-tools/typespec-java/empty-name: Name from TCGC is empty.
08:04:36.616 cmderr W 	[Java-Sdk-Automation] /mnt/vss/_work/1/s/azure-sdk-for-java/sdk/machinelearningservices/azure-resourcemanager-machinelearningservices/TempTypeSpecFiles/MachineLearningServices.Management/CodeVersion.tsp:240:5 - error @azure-tools/typespec-java/empty-name: Name from TCGC is empty.
08:04:36.616 cmderr W 	[Java-Sdk-Automation] /mnt/vss/_work/1/s/azure-sdk-for-java/sdk/machinelearningservices/azure-resourcemanager-machinelearningservices/TempTypeSpecFiles/MachineLearningServices.Management/ComponentVersion.tsp:223:5 - error @azure-tools/typespec-java/empty-name: Name from TCGC is empty.
08:04:36.616 cmderr W 	[Java-Sdk-Automation] /mnt/vss/_work/1/s/azure-sdk-for-java/sdk/machinelearningservices/azure-resourcemanager-machinelearningservices/TempTypeSpecFiles/MachineLearningServices.Management/DataVersionBase.tsp:287:5 - error @azure-tools/typespec-java/empty-name: Name from TCGC is empty.
08:04:36.616 cmderr W 	[Java-Sdk-Automation] /mnt/vss/_work/1/s/azure-sdk-for-java/sdk/machinelearningservices/azure-resourcemanager-machinelearningservices/TempTypeSpecFiles/MachineLearningServices.Management/EnvironmentVersion.tsp:229:5 - error @azure-tools/typespec-java/empty-name: Name from TCGC is empty.
08:04:36.616 cmderr W 	[Java-Sdk-Automation] /mnt/vss/_work/1/s/azure-sdk-for-java/sdk/machinelearningservices/azure-resourcemanager-machinelearningservices/TempTypeSpecFiles/MachineLearningServices.Management/ModelVersion.tsp:300:5 - error @azure-tools/typespec-java/empty-name: Name from TCGC is empty.
08:04:36.616 cmderr W 	[Java-Sdk-Automation] /mnt/vss/_work/1/s/azure-sdk-for-java/sdk/machinelearningservices/azure-resourcemanager-machinelearningservices/TempTypeSpecFiles/MachineLearningServices.Management/OnlineEndpoint.tsp:161:16 - error @azure-tools/typespec-java/empty-name: Name from TCGC is empty.
08:04:36.618 cmderr W 	[Java-Sdk-Automation] /mnt/vss/_work/1/s/azure-sdk-for-java/sdk/machinelearningservices/azure-resourcemanager-machinelearningservices/TempTypeSpecFiles/MachineLearningServices.Management/JobBase.tsp:105:16 - error @azure-tools/typespec-java/empty-name: Name from TCGC is empty.
08:04:36.618 cmderr W 	[Java-Sdk-Automation] /mnt/vss/_work/1/s/azure-sdk-for-java/sdk/machinelearningservices/azure-resourcemanager-machinelearningservices/TempTypeSpecFiles/MachineLearningServices.Management/RaiBlocklistPropertiesBasicResource.tsp:93:5 - error @azure-tools/typespec-java/empty-name: Name from TCGC is empty.
```

weidongxu-microsoft replied:
use this as example of the fix
https://github.com/Azure/azure-rest-api-specs/pull/36687#discussion_r2591330909

mcgallan replied:
> still have one not fixed?
> 
> ```
> 03:21:29.150 cmderr W 	[Java-Sdk-Automation] /mnt/vss/_work/1/s/azure-sdk-for-java/sdk/machinelearningservices/azure-resourcemanager-machinelearningservices/TempTypeSpecFiles/MachineLearningServices.Management/RaiBlocklistPropertiesBasicResource.tsp:93:5 - error @azure-tools/typespec-java/empty-name: Name from TCGC is empty.
> ```

deleteBulk has been fixed

weidongxu-microsoft replied:
@mcgallan

It is line 93, that should be `addBulk` op?

Do double check, its FinalResult may not be a void.

mcgallan committed:
Merge branch 'main' into Convert-MachineLearningServices-preview-10-01

mcgallan committed:
fix void finalresult

mcgallan committed:
fix deleteBulk

weidongxu-microsoft committed:
Update tspconfig.yaml

msyyc commented on specification/machinelearningservices/MachineLearningServices.Management/tspconfig.yaml:
```diff
@@ -0,0 +1,45 @@
+    emit-lro-options: "all"
+    examples-dir: "{project-root}/examples"
+  "@azure-tools/typespec-python":
+    emitter-output-dir: "{output-dir}/{service-dir}/azure-mgmt-machinelearningservices"
```
```suggestion
    emitter-output-dir: "{output-dir}/machinelearning/azure-mgmt-machinelearningservices"
```

msyyc committed:
Update specification/machinelearningservices/MachineLearningServices.Management/tspconfig.yaml

v-jiaodi committed:
Update tspconfig.yaml for Machine Learning services

mcgallan committed:
add void finalresult for addBulk

v-huizhu2 committed:
refactor(spec): update Java client naming for blob reference SAS operations

- Rename GetBlobReferenceSASRequestDto to GetBlobReferenceSasRequestDto for Java
- Rename GetBlobReferenceSASResponseDto to GetBlobReferenceSasResponseDto for Java
- Rename SASCredential to SasCredential for Java
- Rename SASCredentialDto to SasCredentialDto for Java
- Rename SASAuthTypeWorkspaceConnectionProperties to SasAuthTypeWorkspaceConnectionProperties for Java
- Rename Registries.getBlobReferenceSAS to getBlobReferenceSas for Java client

weidongxu-microsoft commented on specification/machinelearningservices/MachineLearningServices.Management/back-compatible.tsp:
```diff
@@ -266,6 +266,15 @@ using Microsoft.MachineLearningServices;
+  "SasAuthTypeWorkspaceConnectionProperties",
+  "java"
+);
+@@clientName(Registries.getBlobReferenceSAS, "getBlobReferenceSas", "java");
```
they need to be in client.tsp

v-huizhu2 committed:
Remove legacy client name annotations

Removed legacy client name annotations for BlobReferenceSAS and SASCredential.

v-huizhu2 committed:
Fix Java casing for SAS-related types

Added Java casing fixes for SAS-related types and operations in client.tsp.

mcgallan committed:
Fix RegistryDataContainers_CreateOrUpdate FinalResult

mcgallan committed:
format

jliusan commented on specification/machinelearningservices/MachineLearningServices.Management/ManagedNetworkSettingsPropertiesBasicResource.tsp:
```diff
@@ -0,0 +1,94 @@
+  @action("batchOutboundRules")
+  @list
+  @post
+  post is ArmResourceActionAsync<
```
paged method Post has no synthesized response type

msyyc committed:
Merge branch 'main' into Convert-MachineLearningServices-preview-10-01

msyyc committed:
fix python issue for api-version

v-huizhu2 committed:
refactor(MachineLearningServices): update Java client naming conventions

- Rename hostname property from 'hostName' to 'hostname'
- Standardize admin username properties to use 'adminUsername' instead of 'adminUserName'
- Update FQDN-related property names to use consistent casing (FQDN, FQDNs)
- Rename various FQDN endpoint related types and properties to follow proper casing
- Apply consistent Java client naming patterns across multiple models

v-jiaodi committed:
Add JavaScript to supported languages in API version

v-huizhu2 committed:
refactor(MachineLearningServices): update FQDN client name annotations

- Replace multiple FQDN-related client name annotations with
  proper Java naming conventions
- Update ExternalFQDNResponse and ExternalFQDNResponseInner
  to use camelCase naming
- Remove redundant FQDN endpoint related annotations

v-huizhu2 committed:
refactor(MachineLearningServices): remove unused client name annotation

Remove the unused ExternalFQDNResponseInner client name annotation
from MachineLearningServices.Management client specification,
keeping only the ExternalFQDNResponse annotation.

v-huizhu2 committed:
Merge branch 'java_mgmt_fix_machinelearningservices' into Convert-MachineLearningServices-preview-10-01

v-huizhu2 committed:
fix(java): correct casing for FQDN-related types in client configuration

Java casing fixes for FQDN-related types to ensure proper naming
conventions are followed in the generated client code.

v-huizhu2 committed:
fix(mlops): correct Java casing for AAD and AKS related types

- Fix casing for AADAuthTypeWorkspaceConnectionProperties to
  AadAuthTypeWorkspaceConnectionProperties
- Update AKS-related type names to proper Java casing:
  AksNetworkingConfiguration -> AKSNetworkingConfiguration
  AksComputeSecrets -> AKSComputeSecrets
  AksComputeSecretsProperties -> AKSComputeSecretsProperties

ChenxiJiang333 committed:
mitigate python breaking changes

ChenxiJiang333 committed:
format python config

v-huizhu2 committed:
fix(mlops): add Java casing fix for CMK encryption property

Add Java client name annotation to properly handle
the enableServiceSideCMKEncryption property in WorkspaceProperties
for consistent casing across the Machine Learning Services Management client.

mcgallan committed:
Merge branch 'main' into Convert-MachineLearningServices-preview-10-01

v-huizhu2 committed:
Merge remote-tracking branch 'upstream/Convert-MachineLearningServices-preview-10-01' into java_mgmt_fix_machinelearningservices

jliusan committed:
add go config

jliusan committed:
add go config

mcgallan committed:
update addBulk finalResult

mcgallan committed:
client format

weidongxu-microsoft committed:
rename for java

weidongxu-microsoft committed:
Merge branch 'main' into Convert-MachineLearningServices-preview-10-01

jliusan committed:
add go clientName

tadelesh committed:
Update client.tsp

jliusan committed:
remove go config

jliusan committed:
edit config

v-jiaodi committed:
Update tspconfig.yaml

jliusan committed:
update go config

jliusan committed:
Merge branch 'Convert-MachineLearningServices-preview-10-01' of https://github.com/mcgallan/azure-rest-api-specs into Convert-MachineLearningServices-preview-10-01

mcgallan committed:
fix

v-jiaodi committed:
update client name

v-jiaodi committed:
disable for js

v-jiaodi committed:
update

v-jiaodi committed:
update

v-jiaodi committed:
disable js

pshao25 committed:
Merge branch 'main' into Convert-MachineLearningServices-preview-10-01

mikeharder commented:
> @mikeharder We always suppress LintDiff as we discussed in #35361.
> 
> That sounds like a pipeline issue? I feel like pipeline issue shouldn't block this PR.

@pshao25: Technically, you should only supress LintDiff, if you are fairly confident the errors are the same before and after your change.  In this case, i think LintDiff may have been failing with errors (but not crashing with OOM) before your change, but failing with OOM after your change.

This should have been a red flag, because it means any changes to the spec after your change, will also fail with OOM.  So, such PRs should not be merged, until issue is understood, and the tooling is fixed.  Otherwise, you are just creating an issue that will need an emergency fix later.

What do you mean by "pipeline issue shouldn't block PR"?  It's never safe to merge a PR, if a check like BreakingChanges or LintDiff is crashing.  You have no idea what bug you are introducing.  Better to wait.
